File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/Core/Object/UniqueNumIdGen.cpp
#include "MyProject.h"
#include "UniqueNumIdGen.h"
#include <limits>
#include "TypeDef.h"
#include "MClassFactory.h"

MY_BEGIN_NAMESPACE(MyNS)

M_IMPLEMENT_AND_REGISTER_CLASS(UniqueNumIdGen, GObject)

UniqueNumIdGen::UniqueNumIdGen(uint baseUniqueId)
{
	this->mBaseId = baseUniqueId;
	this->mCurId = this->mBaseId;
}

uint UniqueNumIdGen::genNewId()
{
	if(std::numeric_limits<uint>::max() == this->mCurId)
	{
		this->mPreIdx = this->mCurId;
		this->mCurId = this->mBaseId;
	}

	this->mPreIdx = this->mCurId;
	this->mCurId++;

	return this->mPreIdx;
}

uint UniqueNumIdGen::getCurId()
{
	return this->mCurId;
}

MY_END_NAMESPACE

File: Project/MyProject/Tool/CSToolKit/Project/ToolKit/ToolKit/Source/Libs/Core/UniqueNumIdGen.cs
namespace SDK.Lib
{
    public class UniqueNumIdGen
    {
        protected uint mBaseId;
        protected uint mPreIdx;
        protected uint mCurId;

        public UniqueNumIdGen(uint baseUniqueId)
        {
            this.mBaseId = baseUniqueId;
            this.mCurId = this.mBaseId;
        }

        public uint genNewId()
        {
            if(uint.MaxValue == this.mCurId)
            {
                this.mPreIdx = this.mCurId;
                this.mCurId = this.mBaseId;
            }

            this.mPreIdx = this.mCurId;
            this.mCurId++;

            return this.mPreIdx;
        }

        public uint getCurId()
        {
            return this.mCurId;
        }
    }
}